blackjax/inference/smc/ess.py

```diff
@@ -0,0 +1,78 @@
+"""All things related to SMC effective sample size"""
+from typing import Callable
+
+import jax
+import jax.numpy as jnp
+
+from blackjax.inference.smc.smc import SMCState
+
+
+def ess(log_weights: jnp.ndarray, log: bool = True):
+    """ Compute the effective sample size from the log-weights in a numerically stable way.
+
+    Parameters
+    ----------
+    log_weights: np.ndarray
+        log-weights of the sample
+    log: bool
+        Compute the log-ESS or the ESS
+
+    Returns
+    -------
+    ess: float
+        The effective sample size
+    """
+
+    log_weights = log_weights - jnp.max(log_weights)
+    w = jnp.exp(log_weights)
+
+    if log:
+        w2: jnp.ndarray = jnp.exp(2 * log_weights)
+        return 2 * jnp.log(w.sum()) - jnp.log(w2.sum())
```

Comment:
This will be stable I believe, I removed the max element from the logw already. Also can't use logaddexp here with these. maybe 2 * logsumexp(logw) - logsumexp(2 logw), but it is wasteful, and less precise, as it does the same thing as I did but doesn't take into account the fact that max(logw) simplifies.